src/borg/repository.py

```diff
@@ -973,8 +973,16 @@ def report_error(msg):
                 report_error(str(err))
                 objects = []
                 if repair:
-                    self.io.recover_segment(segment, filename)
+                    self.io.recover_segment(segment, filepath)
                     objects = list(self.io.iter_objects(segment))
+            except OSError as err:
+                if err.errno == errno.EIO:
+                    report_error("I/O error when reading segment file {}".format(filepath))
+                    objects = []
+                    if repair:
+                        os.remove(filepath)
+                else:
+                    raise
```

Comment:
Hmm, some questions / issues:

1. report_error() is called differently here than in the exception handler directly above.
the information from `err` **must** be included in the output message to not lose information.
that might or not include the information which you have in your message, please check.
and both exception handlers must behave / output in a consistent way.

2. the integrityerror exception handler tries to recover as much as possible from the segment file that had such an error inside. there are ususally multiple objects in a segment file and only one of them might have issues and we can rescue all other ones (see recover_segments).

your new OSError handler now assumes that one EIO is enough to just kill the whole segment file.
are there cases where an I/O error might only affect a file partially and we maybe could recover other parts of same file?

Reply:
> the information from err must be included in the output message to not lose information.

please see the example traceback that shows less information than the proposed message. afaik the exception instance contains the message, the error number and the filepath, which are all included in the message.

could i use `logger.exception` before calling `report_error`?

> the integrityerror exception handler tries to recover as much as possible from the segment file that had such an error inside.

> your new OSError handler now assumes that one EIO is enough to just kill the whole segment file.

my understanding is that in my case the segment file can't be read at all by the os, because the btrfs kernel module detects a checksum mismatch of an inode. hence i would assume parts of it can't be recovered.

> are there cases where an I/O error might only affect a file partially and we maybe could recover other parts of same file?

oh wow, i just realized how big these segment files are, so yes. i guess that would be w/ `LoggedIO.recover_segment`, right? i'll look into that, maybe even throwing an `IntegrityError` further down the code path. but i would avoid creating further execution frames within loops.